src/hotspot/share/oops/klass.inline.hpp

```diff
@@ -75,4 +76,72 @@ inline ByteSize Klass::vtable_start_offset() {
   return in_ByteSize(InstanceKlass::header_size() * wordSize);
 }
 
+// subtype check: true if is_subclass_of, or if k is interface and receiver implements it
+inline bool Klass::is_subtype_of(Klass* k) const {
+  assert(secondary_supers() != nullptr, "must be");
+  const juint off = k->super_check_offset();
+  const juint secondary_offset = in_bytes(secondary_super_cache_offset());
+  if (off == secondary_offset) {
+    return search_secondary_supers(k);
+  } else {
+    Klass* sup = *(Klass**)( (address)this + off );
+    return (sup == k);
+  }
+}
+
+// Hashed search for secondary super k.
+inline bool Klass::lookup_secondary_supers_table(Klass* k) const {
+  uintx bitmap = _secondary_supers_bitmap;
+
+  constexpr int highest_bit_number = SECONDARY_SUPERS_TABLE_SIZE - 1;
```

Comment:
It looks like this needs to be the number of bits in a uintx in order to shift out the high bits, which matches the current definition of SECONDARY_SUPERS_TABLE_SIZE.  But I'm wondering if we should separate the two, so that we could, for example, temporarily reduce SECONDARY_SUPERS_TABLE_SIZE to a smaller size for stress testing without the algorithm breaking.  So highest_bit_number here and in the assembly code would be something like sizeof(uintx) * BitsPerByte - 1.

Reply:
Yes, it is the number of bits in the bitmap. Always 64 up 'til now. The fundamental nature of the algorithm is such that there is an occupancy bit for each possible slot. I could assert that here.

Severe stress testing is supported by `StressSecondarySupers`, which uses a terrible hash function. I think that option tests all of the code paths.